server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java

```diff
@@ -81,20 +74,38 @@ private IndexInput fetchOriginBlob(BlobFetchRequest blobFetchRequest) throws IOE
             return cachedIndexInput;
         });
 
-        if (Objects.isNull(origin)) {
-            // origin is not in file cache, download origin
-
-            // open new origin
-            IndexInput downloaded = downloadBlockLocally(blobFetchRequest);
-
-            // refcount = 0 at the beginning
-            FileCachedIndexInput newOrigin = new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), downloaded);
+        if (origin == null) {
```

Comment:
We do not keep the `IndexInput` open always. In case of cache restore, we insert a `ClosedIndexInput` instance which just has the length of the index input to keep track of capacity. We surely do need a mechanism to keep track of `isClosed` mechanism if the file hasn't been completely eliminated from local disk.

Reply:
>  We surely do need a mechanism to keep track of isClosed mechanism if the file hasn't been completely eliminated from local disk.

Correct, those should be removed, right? And `computeIfAbsent` would work as expected

Reply:
Is your suggestion on the lines of - 

```
// Do download/network fetches with this block
IndexInput origin = fileCache.computeIfAbsent(key, (key) -> downloadBlockLocally(...));
// Once we have the block, either downloaded or on disk, do the following?
if (cachedIndexInput.isClosed()) {
}

```

Reply:
> > We surely do need a mechanism to keep track of isClosed mechanism if the file hasn't been completely eliminated from local disk.
> 
> Correct, those should be removed, right? And `computeIfAbsent` would work as expected

@reta We still need to handle the closed IndexInput case. On startup, we populate the cache with placeholder IndexInput entries and defer creating the real IndexInput entry that actually opens a handle to the file on disk until it is actually needed to be used.